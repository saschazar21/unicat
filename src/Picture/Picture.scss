@import '../__styles__/variables';

.crop {
  position: relative;
  max-height: 100%;
  max-width: 100%;
  margin: 0;
  padding: 0;
  overflow: hidden;
}

.wrapper {
  transition: transition(opacity);
  display: block;
  overflow: hidden;
  height: auto;
  width: 100%;
  opacity: 0;

  .crop & {
    position: absolute;
    object-fit: cover;
    height: 100%;
    width: 100%;
    top: 0;
    left: 0;
  }
}

.loaded {
  opacity: 1;
}

.round {
  border-radius: borderRadius(-1);
}

.rounded {
  border-radius: borderRadius();
}
