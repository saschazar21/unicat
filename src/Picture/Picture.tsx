import React, { Component } from 'react';
import classnames from 'classnames';

import styles from './Picture.scss';

declare module 'react' {
  interface ImgHTMLAttributes<T> {
    loading?: 'lazy' | 'eager' | 'auto';
  }
}

export interface SourceSet {
  url: string;
  width: number;
}

export interface PictureProps {
  className?: string;
  crop?: boolean;
  description: string;
  height?: string;
  lazyload?: boolean;
  loading?: 'lazy' | 'eager' | 'auto';
  media?: string;
  onLoad?: () => void;
  placeholder?: string;
  srcset?: SourceSet[];
  src: string;
  shape?: 'round' | 'rounded' | 'default';
  width?: string;
}

export default class Picture extends Component<PictureProps> {
  private nativeLazyLoading = 'loading' in HTMLImageElement.prototype;

  static defaultProps = {
    lazyload: false,
    loading: 'auto',
    shape: 'default',
    srcset: [],
  };

  state = {
    loaded: false,
  };

  handleOnLoad = (): void => {
    const { onLoad } = this.props;

    this.setState({ loaded: true });

    if (onLoad) {
      onLoad();
    }
  };

  public render(): JSX.Element {
    const {
      lazyload: defaultLazyLoad,
      loading: defaultLoading,
      shape: defaultShape,
    } = Picture.defaultProps;

    const {
      className: customClassName,
      crop,
      description,
      lazyload = defaultLazyLoad,
      loading = defaultLoading,
      media,
      placeholder,
      shape = defaultShape,
      src,
      srcset,
      ...other
    } = this.props;

    const { loaded } = this.state;

    const srcSet = Array.isArray(srcset)
      ? srcset.map(({ url, width }) => `${url} ${width}w`).join(', ')
      : null;

    const className = classnames(
      styles.wrapper,
      { [styles.shape]: shape !== defaultShape, [styles.loaded]: loaded },
      customClassName
    );

    const props = {
      ...other,
      className,
      alt: description,
      media,
      onLoad: this.handleOnLoad,
    };

    const { height = '64px', width = '64px' } = this.props;

    const img = React.createElement(
      'img',
      Object.assign(
        {},
        props,
        { src, srcSet },
        lazyload && !this.nativeLazyLoading
          ? {
              'data-srcset': srcSet,
              'data-src': src,
              src: placeholder,
              srcSet: null,
            }
          : null,
        lazyload || this.nativeLazyLoading ? { loading } : null
      )
    );

    return crop ? (
      <figure className={styles.crop} style={{ height, width }}>
        {img}
      </figure>
    ) : (
      img
    );
  }
}
