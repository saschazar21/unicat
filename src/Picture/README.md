Renders a responsive image using a given `srcset` property.

```jsx
const srcset = [
  {
    url: 'https://placehold.it/2500x2000.jpg',
    width: 2448,
  },
  {
    url: 'https://placehold.it/1920x1080.jpg',
    width: 1080,
  },
  {
    url: 'https://placehold.it/640x480.jpg',
    width: 400,
  },
  {
    url: 'https://placehold.it/320x240.jpg',
    width: 200,
  },
];

<Picture
  srcset={srcset}
  description="A placeholder image showing the currently used resolution"
  src="https://placehold.it/160x120.jpg"
/>;
```

Furthermore, it's possible to crop an image (useful for keeping certain formats, e.g. avatars, etc...)

```jsx
const srcset = [
  {
    url: 'https://placehold.it/2500x2000.jpg',
    width: 2448,
  },
  {
    url: 'https://placehold.it/1920x1080.jpg',
    width: 1080,
  },
  {
    url: 'https://placehold.it/640x480.jpg',
    width: 400,
  },
  {
    url: 'https://placehold.it/320x240.jpg',
    width: 200,
  },
];

<Picture
  crop
  srcset={srcset}
  height="480px"
  width="640px"
  description="A placeholder image showing the currently used resolution"
  src="https://placehold.it/160x120.jpg"
/>;
```
