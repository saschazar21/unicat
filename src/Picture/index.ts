export { default, RawPicture } from './Picture';
