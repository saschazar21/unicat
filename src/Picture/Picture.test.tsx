import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { RawPicture as Picture } from './Picture';

describe('Picture', () => {
  it('renders', () => {
    const wrapper = shallow(
      <Picture description="A sample test image" src="//placehold.it/200x200" />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders a lazy-load variant', () => {
    const wrapper = shallow(
      <Picture
        description="Test"
        src="//placehold.it/200x200"
        lazyload
        srcset={[
          {
            url: '//placehold.it/400x400',
            width: 400,
          },
          {
            url: '//placehold.it/600x600',
            width: 600,
          },
        ]}
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('img[loading]')).toHaveLength(1);
  });
});
