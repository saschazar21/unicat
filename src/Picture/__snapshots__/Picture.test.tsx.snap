// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picture renders 1`] = `
<img
  alt="A sample test image"
  className="wrapper"
  onLoad={[Function]}
  src="//placehold.it/200x200"
  srcSet=""
/>
`;

exports[`Picture renders a lazy-load variant 1`] = `
<img
  alt="Test"
  className="wrapper"
  data-src="//placehold.it/200x200"
  data-srcset="//placehold.it/400x400 400w, //placehold.it/600x600 600w"
  loading="auto"
  onLoad={[Function]}
  srcSet={null}
/>
`;
